{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Workflows for biomedical data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Welcome to the course Computational Workflows for Biomedical Data. Over the next two weeks, you will learn how to leverage nf-core pipelines to analyze biomedical data and gain hands-on experience in creating your own pipelines, with a strong emphasis on Nextflow and nf-core.\n",
    "\n",
    "Course Structure:\n",
    "\n",
    "- Week 1: You will use a variety of nf-core pipelines to analyze a publicly available biomedical study.\n",
    "- Week 2: We will shift focus to learning the basics of Nextflow, enabling you to design and implement your own computational workflows.<br>\n",
    "- Final Project: The last couple of days, you will apply your knowledge to create a custom pipeline for analyzing biomedical data using Nextflow and the nf-core template."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basics\n",
    "\n",
    "If you have not installed all required software, please do so now asap!\n",
    "\n",
    "\n",
    "If you already installed all software, please go on and start answering the questions in this notebook. If you have any questions, don't hesitate to approach us."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. What is nf-core?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "A global community collaborating to build open-source Nextflow components and pipelines.\n",
    "All nf-core code is community owned.\n",
    "Everyone is welcome to use, contribute to, and help maintain nf-core.\n",
    "\n",
    "Quelle: https://nf-co.re/about, https://nf-co.re/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. How many pipelines are there currently in nf-core?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are currently 139 pipelines available as part of nf-core.\n",
    "\n",
    "Quelle: https://nf-co.re/pipelines/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Are there any non-bioinformatic pipelines in nf-core?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The most pipelines seem to be bioinformatic pipelines, however for example rangeland seem to be not bioinformatic in a classical way. It is a geographical best-practice analysis pipeline for remotely sensed imagery. The pipeline processes satellite imagery alongside auxiliary data in multiple steps to arrive at a set of trend files related to land-cover changes.\n",
    "\n",
    "Quelle: https://nf-co.re/rangeland/1.0.0/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Let's go back a couple of steps. What is a pipeline and what do we use it for?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pipeline\n",
    "\n",
    "It is a set of ordered computational steps that take data as input, perform some specific analyses (e.g. alignment, quality control, ...), and produce results as output. Each step within a pipeline is usually reusable, and its ensures that they are connected in the correct order with the right dependencies and resources.\n",
    "\n",
    "What do we use it for?\n",
    "\n",
    "Instead of manually running each tool and handling intermediate files, the pipeline runs everything in the right order. With nf-core pipelines the version of each individual tool used/step done is version-controlled, tested, and containerized, so the same analysis can be reproduced anywhere.\n",
    "They can be run on any laptop, cluster, or cloud without rewriting code and follow strict guidelines for best practices, making them reliable and easier to share.\n",
    "\n",
    "Quelle: input lecture"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Why do you think nf-core adheres to strict guidelines?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nf-core follows strict guidelines because it wants all pipelines to follow the FAIR principles. It wants pipelines to be reproducible, consistent, maintainable (follow the same coding style), trustworthy (quality standards) and portable & scalable (as mentioned above).\n",
    "\n",
    "QUelle: https://nf-co.re/docs/guidelines/pipelines/overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "6. What are the main features of nf-core pipelines?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nf-core provides fully featured pipelines:\n",
    "\n",
    "- Documentation covering installation, usage, and description of output files.\n",
    "- Release of pipelines with tag to stable version.\n",
    "- Open Source: licenced unter the MIT licence.\n",
    "- CI Testing: Uses continuous-integration testing for changed made to pipelines.\n",
    "- Pipelines are ultra-portable (can run everywhere: laptop, cluster, ...)\n",
    "- Packaged Software: Also dependencies are downloaded and handeled automatically.\n",
    "\n",
    "Quelle: https://nf-co.re/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "## Let's start using the pipelines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Find the nf-core pipeline used to measure differential abundance of genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "nf-core/differentialabundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/chrissi/BioPrak/computational-workflows-2025/notebooks/day_01\n"
     ]
    }
   ],
   "source": [
    "# run the pipeline in a cell \n",
    "# to run bash in jupyter notebooks, simply use ! before the command\n",
    "# e.g.\n",
    "\n",
    "!pwd\n",
    "\n",
    "\n",
    "# For the tasks in the first week, please use the command line to run your commands \n",
    "# and simply paste the commands you used in the respective cells!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m\u001b[38;5;232m\u001b[48;5;43m N E X T F L O W \u001b[0;2m  ~  \u001b[mversion 25.04.7\u001b[m\n",
      "\u001b[K\n",
      "Pulling nextflow-io/... ...\n",
      "WARN: Cannot read project manifest -- Cause: Remote resource not found: https://api.github.com/repos/nextflow-io/.../contents/nextflow.config\n",
      "Remote resource not found: https://api.github.com/repos/nextflow-io/.../contents/main.nf\n"
     ]
    }
   ],
   "source": [
    "# run the pipeline in the test profile using docker containers\n",
    "# make sure to specify the version you want to use (use the latest one)\n",
    "\n",
    "\n",
    "!nextflow run nf-core/differentialabundance -r 1.5.0 -profile test,docker --outdir './results'\n",
    "\n",
    "\"\"\"\n",
    "N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [sharp_mclean] DSL2 - revision: 3dd360fed0 [1.5.0]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : 1.5.0\n",
    "  runName                     : sharp_mclean\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /home/chrissi/BioPrak/computational-workflows-2025\n",
    "  workDir                     : /home/chrissi/BioPrak/computational-workflows-2025/work\n",
    "  projectDir                  : /home/chrissi/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : chrissi\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : ./results\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "executor >  local (2)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "executor >  local (2)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "executor >  local (11)\n",
    "executor >  local (12)\n",
    "executor >  local (12)\n",
    "executor >  local (12)\n",
    "executor >  local (13)\n",
    "executor >  local (13)\n",
    "executor >  local (13)\n",
    "executor >  local (14)\n",
    "executor >  local (15)\n",
    "executor >  local (16)\n",
    "executor >  local (17)\n",
    "executor >  local (17)\n",
    "executor >  local (18)\n",
    "executor >  local (19)\n",
    "executor >  local (20)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "executor >  local (20)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "executor >  local (21)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "executor >  local (21)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "executor >  local (21)\n",
    "[76/1d597b] NFC…UNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz) [100%] 1 of 1 ✔\n",
    "[5b/3cb7db] NFC…_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GTF_TO_TABLE (Mus_musculus) [100%] 1 of 1 ✔\n",
    "[e0/f137aa] NFC…IALABUNDANCE:DIFFERENTIALABUNDANCE:VALIDATOR (SRP254919.samplesheet.csv) [100%] 1 of 1 ✔\n",
    "[cc/9e40a9] NFC…TIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_MATRIXFILTER ([id:SRP254919]) [100%] 1 of 1 ✔\n",
    "[c8/ecf40e] NFC…y_hND6_, variable:treatment, reference:mCherry, target:hND6, blocking:]) [100%] 1 of 1 ✔\n",
    "[48/f743fb] NFC…able:treatment, reference:mCherry, target:hND6, blocking:sample_number]) [100%] 2 of 2 ✔\n",
    "[c9/b0f577] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:FILTER_DIFFTABLE (2)      [100%] 2 of 2 ✔\n",
    "[06/e2cf76] NFC…:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEAGCT (treatment_mCherry_hND6_) [100%] 1 of 1 ✔\n",
    "[3f/4830e9] NFC…ABUNDANCE:CUSTOM_TABULARTOGSEACLS (treatment_mCherry_hND6_sample_number) [100%] 2 of 2 ✔\n",
    "[aa/1cf456] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:TABULAR_TO_GSEA_CHIP (1)  [100%] 1 of 1 ✔\n",
    "[3d/c604f2] NFC…E:DIFFERENTIALABUNDANCE:GSEA_GSEA (treatment_mCherry_hND6_sample_number) [100%] 2 of 2 ✔\n",
    "[fd/4f2e71] NFC…DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:PLOT_EXPLORATORY (treatment) [100%] 1 of 1 ✔\n",
    "[ba/48022f] NFC…ENTIALABUNDANCE:PLOT_DIFFERENTIAL (treatment_mCherry_hND6_sample_number) [100%] 2 of 2 ✔\n",
    "[fe/658fd7] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:SHINYNGS_APP (SRP254919)  [100%] 1 of 1 ✔\n",
    "[d7/39ae51] NFC…IFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:RMARKDOWNNOTEBOOK (SRP254919) [100%] 1 of 1 ✔\n",
    "[f7/d76cac] NFC…FFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:MAKE_REPORT_BUNDLE (SRP254919) [100%] 1 of 1 ✔\n",
    "Completed at: 29-Sep-2025 12:14:59\n",
    "Duration    : 9m 47s\n",
    "CPU hours   : (a few seconds)\n",
    "Succeeded   : 21\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# repeat the run. What did change?\n",
    "\n",
    "!nextflow run nf-core/differentialabundance -r 1.5.0 -profile test,docker --outdir './results_repeat'\n",
    "\n",
    "\"\"\"\n",
    " N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [clever_baekeland] DSL2 - revision: 3dd360fed0 [1.5.0]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : 1.5.0\n",
    "  runName                     : clever_baekeland\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /home/chrissi/BioPrak/computational-workflows-2025\n",
    "  workDir                     : /home/chrissi/BioPrak/computational-workflows-2025/work\n",
    "  projectDir                  : /home/chrissi/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : chrissi\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : ./results_repeat\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "executor >  local (1)\n",
    "executor >  local (2)\n",
    "executor >  local (2)\n",
    "executor >  local (3)\n",
    "executor >  local (4)\n",
    "executor >  local (5)\n",
    "executor >  local (11)\n",
    "executor >  local (12)\n",
    "executor >  local (12)\n",
    "executor >  local (13)\n",
    "executor >  local (13)\n",
    "executor >  local (14)\n",
    "executor >  local (15)\n",
    "executor >  local (15)\n",
    "executor >  local (16)\n",
    "executor >  local (17)\n",
    "executor >  local (17)\n",
    "executor >  local (18)\n",
    "executor >  local (18)\n",
    "executor >  local (19)\n",
    "executor >  local (20)\n",
    "[dd/4fafb2] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "executor >  local (20)\n",
    "[dd/4fafb2] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "executor >  local (20)\n",
    "[dd/4fafb2] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "executor >  local (21)\n",
    "[dd/4fafb2] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "executor >  local (21)\n",
    "[dd/4fafb2] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "executor >  local (21)\n",
    "[dd/4fafb2] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "[06/79416a] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GTF_TO_TABLE (Mus_musculus)                                 [100%] 1 of 1 ✔\n",
    "[7b/604907] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:VALIDATOR (SRP254919.samplesheet.csv)                       [100%] 1 of 1 ✔\n",
    "[74/600288] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_MATRIXFILTER ([id:SRP254919])                        [100%] 1 of 1 ✔\n",
    "[21/f4e0ba] NFC…:DESEQ2_NORM ([id:treatment_mCherry_hND6_, variable:treatment, reference:mCherry, target:hND6, blocking:]) [100%] 1 of 1 ✔\n",
    "[1b/73eb5a] NFC…DIFFERENTIAL ([id:treatment_mCherry_hND6_, variable:treatment, reference:mCherry, target:hND6, blocking:]) [100%] 2 of 2 ✔\n",
    "[b6/515621] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:FILTER_DIFFTABLE (2)                                        [100%] 2 of 2 ✔\n",
    "[1e/148fa4] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEAGCT (treatment_mCherry_hND6_)           [100%] 1 of 1 ✔\n",
    "[96/807d5f] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEACLS (treatment_mCherry_hND6_)           [100%] 2 of 2 ✔\n",
    "[50/ee8741] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:TABULAR_TO_GSEA_CHIP (1)                                    [100%] 1 of 1 ✔\n",
    "[7d/413110] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GSEA_GSEA (treatment_mCherry_hND6_sample_number)            [100%] 2 of 2 ✔\n",
    "[07/f8caf6] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:PLOT_EXPLORATORY (treatment)                                [100%] 1 of 1 ✔\n",
    "[60/83e1da] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:PLOT_DIFFERENTIAL (treatment_mCherry_hND6_)                 [100%] 2 of 2 ✔\n",
    "[15/ebaedd] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:SHINYNGS_APP (SRP254919)                                    [100%] 1 of 1 ✔\n",
    "[ea/896cdb] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:RMARKDOWNNOTEBOOK (SRP254919)                               [100%] 1 of 1 ✔\n",
    "[22/665bbd] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:MAKE_REPORT_BUNDLE (SRP254919)                              [100%] 1 of 1 ✔\n",
    "Completed at: 29-Sep-2025 12:22:35\n",
    "Duration    : 3m 28s\n",
    "CPU hours   : (a few seconds)\n",
    "Succeeded   : 21\n",
    "\"\"\"\n",
    "\n",
    "!nextflow run nf-core/differentialabundance -r 1.5.0 -profile test,docker --outdir './results_repeat'\n",
    "\n",
    "\"\"\"\n",
    " N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [nice_bartik] DSL2 - revision: 3dd360fed0 [1.5.0]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : 1.5.0\n",
    "  runName                     : nice_bartik\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /home/chrissi/BioPrak/computational-workflows-2025\n",
    "  workDir                     : /home/chrissi/BioPrak/computational-workflows-2025/work\n",
    "  projectDir                  : /home/chrissi/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : chrissi\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : ./results_repeat\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "executor >  local (1)\n",
    "executor >  local (2)\n",
    "executor >  local (21)\n",
    "[1b/38f987] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GUNZIP_GTF (Mus_musculus.GRCm38.81.gtf.gz)                  [100%] 1 of 1 ✔\n",
    "[00/c9ed4e] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GTF_TO_TABLE (Mus_musculus)                                 [100%] 1 of 1 ✔\n",
    "[44/7a9a45] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:VALIDATOR (SRP254919.samplesheet.csv)                       [100%] 1 of 1 ✔\n",
    "[8f/296b51] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_MATRIXFILTER ([id:SRP254919])                        [100%] 1 of 1 ✔\n",
    "[96/d47436] NFC…:DESEQ2_NORM ([id:treatment_mCherry_hND6_, variable:treatment, reference:mCherry, target:hND6, blocking:]) [100%] 1 of 1 ✔\n",
    "[21/dda798] NFC…t_mCherry_hND6_sample_number, variable:treatment, reference:mCherry, target:hND6, blocking:sample_number]) [100%] 2 of 2 ✔\n",
    "[da/d52e49] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:FILTER_DIFFTABLE (2)                                        [100%] 2 of 2 ✔\n",
    "[d9/dd4f2c] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEAGCT (treatment_mCherry_hND6_)           [100%] 1 of 1 ✔\n",
    "[4f/743cb9] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEACLS (treatment_mCherry_hND6_)           [100%] 2 of 2 ✔\n",
    "[19/76051e] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:TABULAR_TO_GSEA_CHIP (1)                                    [100%] 1 of 1 ✔\n",
    "[5f/d69546] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:GSEA_GSEA (treatment_mCherry_hND6_sample_number)            [100%] 2 of 2 ✔\n",
    "[d0/620c07] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:PLOT_EXPLORATORY (treatment)                                [100%] 1 of 1 ✔\n",
    "[32/4f55ca] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:PLOT_DIFFERENTIAL (treatment_mCherry_hND6_sample_number)    [100%] 2 of 2 ✔\n",
    "[1a/844a63] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:SHINYNGS_APP (SRP254919)                                    [100%] 1 of 1 ✔\n",
    "[6d/5c1411] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:RMARKDOWNNOTEBOOK (SRP254919)                               [100%] 1 of 1 ✔\n",
    "[28/99bb0b] NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:MAKE_REPORT_BUNDLE (SRP254919)                              [100%] 1 of 1 ✔\n",
    "Completed at: 29-Sep-2025 12:36:12\n",
    "Duration    : 3m 17s\n",
    "CPU hours   : 0.1\n",
    "Succeeded   : 21\n",
    "\"\"\"\n",
    "\n",
    "# Runtime decreased from 9min 47s to 3min 28s minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now set -resume to the command. What did change?\n",
    "\n",
    "!nextflow run nf-core/differentialabundance -r 1.5.0 -profile test,docker --outdir './results_resume' -resume"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check out the current directory. Next to the outdir you specified, what else has changed?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete the work directory and run the pipeline again using -resume. What did change?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What changed?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lets look at the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is differential abundance analysis?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Give the most important plots from the report:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "workflows",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
